### DISCUSSION TITLE
.NET 7 Preview: Reflection, COM interop regressions

### DISCUSSION THREAD

**Discussion by ClearScriptLib** [dotnet/runtime#74977] [Q&A]
Greetings!

As a bridge between .NET and external script engines, our project makes heavy use of reflection, COM interop, and the dynamic infrastructure.

Today we started testing on .NET 7 (currently Preview 7), and we're seeing numerous regressions in these areas. After encountering no major issues on .NET Core 3.1, .NET 5, or .NET 6, our code can't get off the ground at all on .NET 7.

Before we start writing up issues, we were wondering if significant breaking changes in these areas are expected in .NET 7.

Thanks!

**Comment by stephentoub** [comment]
cc: @AaronRobinsonMSFT

**Reply by danmoseley** [reply]
@dotnet/interop-contrib

**Comment by AaronRobinsonMSFT** [comment]
> we're seeing numerous regressions in these areas

@ClearScriptLib Oh boy. You're scaring me. It would seem your product is using all the of the sensitive bits. The short answer is no, nothing has intentionally changed to warrant substantial regressions. The COM interop sub-system has remained untouched to the best of my knowledge, however Reflection is the sticky bit.`IUnknown`-based COM interop shouldn't have any changes since .NET 6. However, since you mentioned scripting that means `IDispatch` is likely in play and that heavily relies upon .NET Reflection, which has been updated in a myriad of ways. We had found a few regressions in Reflection and fixed them, but there are possibly others.

With all that in mind, it would be helpful to understand what "regressions" means. Is this performance or functional regressions? Do you have some simple examples we could look at?

/cc @steveharter

**Reply by ClearScriptLib** [reply]
Hi @AaronRobinsonMSFT,

Wow, thanks for the quick response!

>Oh boy. You're scaring me.

Sorry, that wasn't our intention. The results of our initial tests on .NET 7 were alarming, but after some tinkering, things definitely look less dire. Again, apologies for any undue stress.

>Is this performance or functional regressions?

We're concerned only with functional regressions.

>Do you have some simple examples we could look at?

Many of our test failures turned out to be downstream effects of a rather trivial regression in [`Type.InvokeMember`](https://docs.microsoft.com/en-us/dotnet/api/system.type.invokemember). Previously, when given `BindingFlags.CreateInstance`, the method allowed the member name argument to be `null`. Now it throws `ArgumentNullException` in that situation. Passing in `string.Empty` fixes that.

A more serious problem is related to COM interop. On Windows, our library supports the legacy script engines JScript and VBScript. On .NET 7, we can instantiate these engines and execute scripts as always. However, if a script attempts to invoke managed code – which it does via .NET's `IDispatch`-`IReflect` interop – the process dies with an access violation.

Interestingly, this always happens at the same place in the script engine code. The stack looks like this:

```
 jscript.dll!JAmsi::JAmsiProcessor()
 jscript.dll!IDispatchExInvokeEx2()
 jscript.dll!IDispatchExInvokeEx()
 jscript.dll!InvokeDispatchEx()
 jscript.dll!VAR::InvokeByDispID()
 jscript.dll!NameTbl::InvokeInternal()
 jscript.dll!VAR::InvokeByDispID()
 jscript.dll!CScriptRuntime::Run(class VAR *)
 jscript.dll!ScrFncObj::PerformCall(class CScriptRuntime *,class VAR *,int,class VAR *,class VAR *,unsigned long)
 jscript.dll!ScrFncObj::CallWithFrameOnStack(class VAR *,int,class VAR *,class VAR *,unsigned long)
 jscript.dll!ScrFncObj::Call()
 jscript.dll!JsEval()
 jscript.dll!NatFncObj::Call()
 jscript.dll!NameTbl::InvokeInternal()
 jscript.dll!VAR::InvokeByDispID()
 jscript.dll!CScriptRuntime::Run(class VAR *)
 jscript.dll!ScrFncObj::PerformCall(class CScriptRuntime *,class VAR *,int,class VAR *,class VAR *,unsigned long)
 jscript.dll!ScrFncObj::CallWithFrameOnStack(class VAR *,int,class VAR *,class VAR *,unsigned long)
 jscript.dll!ScrFncObj::Call()
 jscript.dll!CSession::Execute()
 jscript.dll!COleScript::ExecutePendingScripts()
 jscript.dll!COleScript::ParseScriptTextCore()
 jscript.dll!COleScript::ParseScriptText(unsigned short const *,unsigned short const *,struct IUnknown *,unsigned short const *,unsigned long,unsigned long,unsigned long,struct tagVARIANT *,struct tagEXCEPINFO *)
```

The access violation occurs here:

```
    54B67FC0  mov         ecx,esi  
    54B67FC2  call        dword ptr [___guard_check_icall_fptr (54BD73E4h)]  
    54B67FC8  call        esi  
--> 54B67FCA  cmp         edi,esp  
    54B67FCC  je          JAmsi::JAmsiProcessor+0C5h (54B67FD5h)  
    54B67FCE  mov         ecx,4  
    54B67FD3  int         29h  
```

If we're to believe the debugger, `esi` holds a value of 2 (definitely not a pointer) at crash time, so the problem is probably with the `call esi` instruction. JScript and VBScript both crash in this manner. Any idea what that `___guard_check_icall_fptr` call does? Anyway, the crash occurs _before_ the native-to-managed transition.

Note that the latter issue is currently blocking our test suite, so we could discover other regressions once we get around it.

Any ideas?

Thank you!

**Reply by AaronRobinsonMSFT** [reply]
> Many of our test failures turned out to be downstream effects of a rather trivial regression in [Type.InvokeMember](https://docs.microsoft.com/en-us/dotnet/api/system.type.invokemember). Previously, when given BindingFlags.CreateInstance, the method allowed the member name argument to be null. Now it throws ArgumentNullException in that situation. Passing in string.Empty fixes that.

I don't recall if we documented this change. @steveharter or @buyaa-n Can you confirm this change was intentional?

> However, if a script attempts to invoke managed code – which it does via .NET's IDispatch-IReflect interop – the process dies with an access violation.

Is there any chance I can get access to a DMP of that crash? This is an area that isn't tested often because it is old and we avoid touching it when possible. I suppose something in the `IDispatch` area could have regressed, but I'd need a DMP or repro to confirm.

> Any idea what that ___guard_check_icall_fptr call does?

That is for Windows [Control Flow Guard](https://docs.microsoft.com/windows/win32/secbp/control-flow-guard).

**Reply by AaronRobinsonMSFT** [reply]
At a minumum it would be helpful to see the precise version of `jscript.dll` being used. From within WinDBG, you can get this by running `lmvmjscript`.

**Reply by stephentoub** [reply]
> I don't recall if we documented this change. @steveharter or @buyaa-n Can you confirm this change was intentional?

I inadvertently caused this when I rolled out usage of `!!`. The analyzer/fixer I used had an issue where it would flag `if (name == null)` cases even if the resulting `!!` usage changed order of operations; I thought I'd caught them all but I apparently missed this one.  I will submit a PR to fix it.

**Reply by ClearScriptLib** [reply]
Hi @AaronRobinsonMSFT,

>Is there any chance I can get access to a DMP of that crash?

Of course: [ClearScriptConsole.zip](https://1drv.ms/u/s!Ag2Yp8AIo2CdbZ-lqB-DIZIf7Zk)

>That is for Windows [Control Flow Guard](https://docs.microsoft.com/windows/win32/secbp/control-flow-guard).

Hmm. Assuming that .NET 7 doesn't install new versions of JScript and VBScript, could it have something to do with [this](https://github.com/dotnet/runtime/issues/59190)? Umm, maybe not, since I can repro this on both x86 and x64?

>At a minumum it would be helpful to see the precise version of jscript.dll being used.

According to VS 2022 17.4.0 Preview 1, that DLL is version 5.812.10240.16384 at C:\Windows\SysWOW64\jscript.dll.

Thanks!

**Reply by ClearScriptLib** [reply]
Hi @AaronRobinsonMSFT,

By switching to x64 and ditching mixed mode, we got a better stack:

```
[Inline Frame] coreclr.dll!InterlockedCompareExchangeT(ITypeInfo * volatile *) Line 4341
[Inline Frame] coreclr.dll!InterlockedCompareExchangeT(ITypeInfo * volatile *) Line 4375
[Inline Frame] coreclr.dll!InterlockedCompareExchangeT(ITypeInfo * volatile *) Line 4406
[Inline Frame] coreclr.dll!ComMethodTable::SetITypeInfo(ITypeInfo * pNew=0x000001efb0d30fe8) Line 3853
coreclr.dll!GetITypeInfoForMT(ComMethodTable * pMT=0x00007ffd575cf440, ITypeInfo * * ppTI=0x0000001d1d17c308) Line 805
coreclr.dll!DispatchEx_GetTypeInfo(IDispatch * pDisp=0x0000001d1d17c258, unsigned int pptinfo=0x0000001d1d17c308, unsigned long) Line 1547
coreclr.dll!DispatchEx_GetTypeInfo_CallBack(void * ptr=0x0000001d1d17c258) Line 1281
coreclr.dll!DispatchEx_GetTypeInfo_Wrapper(IDispatchEx * pDisp=0x00000230477a6740, unsigned int itinfo=0x00000000, unsigned long lcid=0x00000000, ITypeInfo * * pptinfo=0x0000001d1d17c308) Line 1306
jscript.dll!JAmsi::JAmsiProcessor()
jscript.dll!VAR::InvokeByDispID()
jscript.dll!NameTbl::InvokeInternal()
jscript.dll!VAR::InvokeByDispID()
jscript.dll!CScriptRuntime::Run(class VAR *)
jscript.dll!ScrFncObj::PerformCall(class CScriptRuntime *,class VAR *,int,class VAR *,class VAR *,unsigned long)
jscript.dll!ScrFncObj::CallWithFrameOnStack(class VAR *,int,class VAR *,class VAR *,unsigned long)
jscript.dll!ScrFncObj::Call()
jscript.dll!JsEval()
jscript.dll!NatFncObj::Call()
jscript.dll!NameTbl::InvokeInternal()
jscript.dll!VAR::InvokeByDispID()
jscript.dll!CScriptRuntime::Run(class VAR *)
jscript.dll!ScrFncObj::PerformCall(class CScriptRuntime *,class VAR *,int,class VAR *,class VAR *,unsigned long)
jscript.dll!ScrFncObj::CallWithFrameOnStack(class VAR *,int,class VAR *,class VAR *,unsigned long)
jscript.dll!ScrFncObj::Call()
jscript.dll!CSession::Execute(class NameTbl *,struct tagVARIANT *,int,struct tagVARIANT *,struct tagVARIANT *,unsigned long)
jscript.dll!COleScript::ExecutePendingScripts()
jscript.dll!COleScript::ParseScriptTextCore()
jscript.dll!COleScript::ParseScriptText(unsigned short const *,unsigned short const *,struct IUnknown *,unsigned short const *,unsigned __int64,unsigned long,unsigned long,struct tagVARIANT *,struct tagEXCEPINFO *)
```

The debugger is pointing at the following source line in ...\src\coreclr\inc\utilcode.h

``` C++
    template <typename T>
    inline T* InterlockedCompareExchangeT(
        T* volatile * destination,
        T*            exchange,
        T*            comparand)
    {
        //STATIC_ASSERT(exchange == 0);
        typedef typename std::remove_const<T>::type * non_const_ptr_t;
-->     return reinterpret_cast<T*>(InterlockedCompareExchangePointer(
            reinterpret_cast<PVOID volatile *>(const_cast<non_const_ptr_t volatile *>(destination)),
            reinterpret_cast<PVOID>(const_cast<non_const_ptr_t>(exchange)),
            reinterpret_cast<PVOID>(const_cast<non_const_ptr_t>(comparand))));
    }
```

Here's what the disassembler has to say:

```
    // Returns a NON-ADDREF'd ITypeInfo.
    HRESULT GetITypeInfoForMT(ComMethodTable *pMT, ITypeInfo **ppTI)
    {
     mov         r11,rsp  
     mov         qword ptr [r11+10h],rbx  
     mov         qword ptr [r11+18h],rsi  
     push        rdi  
     sub         rsp,20h  
        CONTRACTL
        {
            NOTHROW;
            GC_TRIGGERS;
            MODE_PREEMPTIVE;
            PRECONDITION(CheckPointer(pMT));
            PRECONDITION(CheckPointer(ppTI));
        }
        CONTRACTL_END;

        HRESULT     hr = S_OK;              // A result.
        ITypeInfo   *pTI;                   // The ITypeInfo.

        pTI = pMT->GetITypeInfo();
     mov         rax,qword ptr [rcx+28h]  
     xor         edi,edi  
     mov         qword ptr [r11+8],rax  
     mov         rsi,rdx  
     mov         rbx,rcx  
     test        rax,rax  

        if (pTI == NULL)
     jne         GetITypeInfoForMT+64h (07FFDB740A954h)  
        {
            MethodTable *pClass = pMT->GetMethodTable();

            hr = GetITypeInfoForEEClass(pClass, &pTI);
     mov         rcx,qword ptr [rcx+8]  
     lea         rdx,[r11+8]  
     xor         r8d,r8d  
     call        GetITypeInfoForEEClass (07FFDB7409A70h)  
     mov         edi,eax  

            if (SUCCEEDED(hr))
     test        eax,eax  
     js          GetITypeInfoForMT+5Fh (07FFDB740A94Fh)  
            {
                pMT->SetITypeInfo(pTI);
     mov         rcx,qword ptr [pTI]  
     xor         eax,eax  
-->  lock cmpxchg qword ptr [rbx+28h],rcx  
     jne         GetITypeInfoForMT+4Fh (07FFDB740A93Fh)  
     call        SafeAddRef (07FFDB711731Ch)  
                SafeReleasePreemp(pTI);
     mov         rcx,qword ptr [pTI]  
     call        SafeReleasePreemp (07FFDB7119290h)  
                pTI = pMT->GetITypeInfo();
     mov         rax,qword ptr [rbx+28h]  
     jmp         GetITypeInfoForMT+64h (07FFDB740A954h)  

            if (SUCCEEDED(hr))
     mov         rax,qword ptr [pTI]  
            }
        }

        *ppTI = pTI;
        return hr;
    }
     mov         rbx,qword ptr [rsp+38h]  
     mov         qword ptr [rsi],rax  
     mov         eax,edi  
     mov         rsi,qword ptr [rsp+40h]  
     add         rsp,20h  
     pop         rdi  
     ret  
```

Just for completeness, `[rbx+28h]` is pointing at two consecutive 64-bit nulls followed by other data.

Please let us know if you need additional information. Thanks!

**Reply by ClearScriptLib** [reply]
One other bit: The debugger indicates that the access violation is due to an invalid _write_.

**Reply by jkotas** [reply]
Looks like a regression from W^X work. Do we need to flip the page to writeable around this write?

cc @janvorli

**Reply by jkotas** [reply]
Opened https://github.com/dotnet/runtime/issues/75035 to have it tracked

**Reply by AaronRobinsonMSFT** [reply]
@ClearScriptLib A fix for writing to executable memory has been submitted for .NET 8, https://github.com/dotnet/runtime/pull/75045. I've also opened one for .NET 7, https://github.com/dotnet/runtime/pull/75046.

**Reply by ClearScriptLib** [reply]
@AaronRobinsonMSFT Wow, thank you for the fast work! Is there any way for us to test against the fix?

**Reply by AaronRobinsonMSFT** [reply]
@ClearScriptLib The easiest way is to build the branch locally and install it - https://github.com/dotnet/runtime/pull/75046. In order to build, the following will build the product and installers.

On Windows:
**64-bit** `build.cmd -subset clr+libs+host+Packs.Product -c Release -arch x64`
**32-bit** `build.cmd -subset clr+libs+host+Packs.Product -c Release -arch x86`

**Reply by ClearScriptLib** [reply]
Thanks @AaronRobinsonMSFT. Does `build.cmd` have any specific pre-reqs? VS version, SDKs, other tools?

**Reply by AaronRobinsonMSFT** [reply]
Visual Studio 2019 or 2022 should work without issue. The only other thing I think that you might need is CMake.

**Reply by jkoritzinsky** [reply]
I think our build only works with VS2022 nowadays. I think we dropped support for building with VS2019.

**Reply by ClearScriptLib** [reply]
Folks,

Thanks for all your help so far.

We've managed to build the runtime (didn't need CMake but needed Python 😉) and verified the fix in a simple test:

```
> dotnet run --fx-version 7.0.0-dev -c Release -f net7.0
-> System.Console.WriteLine('Hello, world!')
Hello, world!
->
```

However, to run our full suite, we need `-f net7.0-windows`. The following doesn't work:

```
> dotnet run --fx-version 7.0.0-dev -c Release -f net7.0-windows
The specified framework 'Microsoft.NETCore.App', version '7.0.0-dev', apply_patches=0, version_compatibility_range=exact cannot roll-forward to the previously referenced version '7.0.0-preview.7.22375.6'.
```

The error message seems to suggest that we need to build and install the Windows support as well. How does one do that?

Thanks again!

**Reply by ClearScriptLib** [reply]
Actually, we might be able to run the bulk of the suite without `net7.0-windows`. In either case, we need to force `dotnet test` to use the locally-built `7.0.0-dev` runtime, but `dotnet test` doesn't seem to support `--fx-version`. Suggestions?

**Reply by ClearScriptLib** [reply]
Update: A hacky solution to be sure, but we've managed to run our full test suite on the patched runtime by renaming directories in `C:\Program Files\dotnet\shared\Microsoft.NETCore.App`. As far as we can tell, the problem is fixed! Thank you!

**Reply by AaronRobinsonMSFT** [reply]
Thank you @ClearScriptLib. Your effort is much appreciated.
